File: assembly-adapter/src/main/java/me/panpf/adapter/pager/AssemblyPagerItemFactory.java
/*
 * Copyright (C) 2017 Peng fei Pan <sky@panpf.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.panpf.adapter.pager;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import me.panpf.adapter.OnClickListener;
import me.panpf.adapter.OnLongClickListener;

public abstract class AssemblyPagerItemFactory<DATA> {

    @Nullable
    private AssemblyPagerAdapter adapter;
    @Nullable
    private PagerClickListenerManager<DATA> clickListenerManager;

    @Nullable
    public AssemblyPagerAdapter getAdapter() {
        return adapter;
    }

    void attachToAdapter(@Nullable AssemblyPagerAdapter adapter) {
        this.adapter = adapter;
    }

    public AssemblyPagerItemFactory<DATA> setOnViewClickListener(@IdRes int viewId, @NonNull OnClickListener<DATA> onClickListener) {
        if (clickListenerManager == null) {
            clickListenerManager = new PagerClickListenerManager<DATA>();
        }
        clickListenerManager.add(viewId, onClickListener);
        return this;
    }

    public AssemblyPagerItemFactory<DATA> setOnItemClickListener(@NonNull OnClickListener<DATA> onClickListener) {
        if (clickListenerManager == null) {
            clickListenerManager = new PagerClickListenerManager<DATA>();
        }
        clickListenerManager.add(onClickListener);
        return this;
    }

    public AssemblyPagerItemFactory<DATA> setOnViewLongClickListener(@IdRes int viewId, @NonNull OnLongClickListener<DATA> onClickListener) {
        if (clickListenerManager == null) {
            clickListenerManager = new PagerClickListenerManager<DATA>();
        }
        clickListenerManager.add(viewId, onClickListener);
        return this;
    }

    public AssemblyPagerItemFactory<DATA> setOnItemLongClickListener(@NonNull OnLongClickListener<DATA> onClickListener) {
        if (clickListenerManager == null) {
            clickListenerManager = new PagerClickListenerManager<DATA>();
        }
        clickListenerManager.add(onClickListener);
        return this;
    }

    protected View dispatchCreateView(@NonNull Context context, @NonNull ViewGroup container, int position, @Nullable DATA data) {
        View itemView = createView(context, container, position, data);

        if (clickListenerManager != null) {
            clickListenerManager.register(this, itemView, position, data);
        }

        return itemView;
    }

    /**
     * 匹配数据
     *
     * @param data 待匹配的数据，通常是使用 instanceof 关键字匹配类型
     * @return 如果返回 true，Adapter 将会使用此 item factory 来处理当前这条数据
     */
    public abstract boolean match(@Nullable Object data);

    /**
     * 创建 View
     */
    public abstract View createView(@NonNull Context context, @NonNull ViewGroup container, int position, @Nullable DATA data);
}

File: assemblyadapter-pager/src/main/java/com/github/panpf/assemblyadapter/pager/AssemblyPagerItemFactory.kt
/*
 * Copyright (C) 2017 Peng fei Pan <sky@panpf.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.assemblyadapter.pager

import android.content.Context
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IdRes
import com.github.panpf.assemblyadapter.OnClickListener
import com.github.panpf.assemblyadapter.OnLongClickListener
import com.github.panpf.assemblyadapter.internal.BaseItemFactory
import com.github.panpf.assemblyadapter.internal.ClickListenerManager

abstract class AssemblyPagerItemFactory<DATA> : BaseItemFactory {

    private var clickListenerManager: ClickListenerManager<DATA>? = null

    fun dispatchCreateView(
        context: Context, container: ViewGroup, position: Int, data: DATA?
    ): View {
        return createView(context, container, position, data).apply {
            registerItemClickListener(this, position, data)
        }
    }

    abstract override fun match(data: Any?): Boolean

    abstract fun createView(
        context: Context, container: ViewGroup, position: Int, data: DATA?
    ): View

    fun setOnViewClickListener(
        @IdRes viewId: Int, onClickListener: OnClickListener<DATA>
    ): AssemblyPagerItemFactory<DATA> {
        getClickListenerManagerOrCreate().add(viewId, onClickListener)
        return this
    }

    fun setOnItemClickListener(onClickListener: OnClickListener<DATA>): AssemblyPagerItemFactory<DATA> {
        getClickListenerManagerOrCreate().add(onClickListener)
        return this
    }

    fun setOnViewLongClickListener(
        @IdRes viewId: Int, onClickListener: OnLongClickListener<DATA>
    ): AssemblyPagerItemFactory<DATA> {
        getClickListenerManagerOrCreate().add(viewId, onClickListener)
        return this
    }

    fun setOnItemLongClickListener(onClickListener: OnLongClickListener<DATA>): AssemblyPagerItemFactory<DATA> {
        getClickListenerManagerOrCreate().add(onClickListener)
        return this
    }

    private fun getClickListenerManagerOrCreate(): ClickListenerManager<DATA> {
        return (clickListenerManager ?: (ClickListenerManager<DATA>().apply {
            this@AssemblyPagerItemFactory.clickListenerManager = this
        }))
    }

    private fun registerItemClickListener(itemView: View, position: Int, data: DATA?) {
        val clickListenerManager = clickListenerManager ?: return
        for (holder in clickListenerManager.holders) {
            if (holder is ClickListenerManager.ClickListenerHolder<*>) {
                @Suppress("UNCHECKED_CAST")
                val clickListenerHolder = holder as ClickListenerManager.ClickListenerHolder<DATA>
                val viewId: Int = clickListenerHolder.viewId
                val targetView: View = if (viewId > 0) {
                    itemView.findViewById(viewId)
                        ?: throw IllegalArgumentException("Not found click bind target view by id $viewId")
                } else {
                    itemView
                }
                targetView.setOnClickListener { view ->
                    clickListenerHolder.listener.onClick(view.context, view, position, data)
                }
            } else if (holder is ClickListenerManager.LongClickListenerHolder<*>) {
                @Suppress("UNCHECKED_CAST")
                val longClickListenerHolder =
                    holder as ClickListenerManager.LongClickListenerHolder<DATA>
                val viewId: Int = longClickListenerHolder.viewId
                val targetView: View = if (viewId > 0) {
                    itemView.findViewById(viewId)
                        ?: throw IllegalArgumentException("Not found click bind target view by id $viewId")
                } else {
                    itemView
                }
                targetView.setOnLongClickListener { view ->
                    longClickListenerHolder.listener.onLongClick(view.context, view, position, data)
                }
            }
        }
    }
}